### PR TITLE
fix: backward compatible modeling of try-with-resource

### PR DESCRIPTION
**Pull request by monperrus**
An iteration about try-with-resources on top of https://github.com/INRIA/spoon/pull/4371.

Before the resource declaration was a CtVariableRead, which is not natural and breaking.

Now a resource is an implicit variable declaration. The main advantage is that it is fully backward compatible (hence we would fix https://ci.inria.fr/sos/job/npefix/)

Plus, I've added a test on pretty-printing.

### CODE REVIEW

**Comment by monperrus** [comment]
the diff is now clean here

File: src/test/java/spoon/test/trycatch/TryCatchTest.java [hunk 1]
```diff
@@ -421,15 +422,18 @@ public void testTryWithVariableAsResource() {
...
         List<CtResource<?>> resources = tryStmt.getResources();
         assertEquals(1, resources.size());
 		final CtResource<?> ctResource = resources.get(0);
-		assertTrue(ctResource instanceof CtVariableRead);
-        assertEquals("resource", ((CtVariableRead<?>) ctResource).getVariable().getSimpleName());
+		assertTrue(ctResource instanceof CtVariable);
+        assertEquals("resource", ((CtVariable<?>) ctResource).getSimpleName());
+
+		// contract: pretty-printing of existing resources works
+		assertEquals("try (resource) {"+System.getProperty("line.separator")+"}", tryStmt.toString());
```
Comment (MartinWitt) [inline src/test/java/spoon/test/trycatch/TryCatchTest.java:436]: You could use the [LinespeparatorExtension](https://github.com/INRIA/spoon/blob/59990fe39101fdb7fcb13d4ad6bbaf0085cb8509/src/test/java/spoon/testing/utils/LineSeperatorExtension.java) here and `\n` in the testcase.

File: src/test/java/spoon/test/trycatch/TryCatchTest.java [hunk 2]
```diff
@@ -421,15 +422,18 @@ public void testTryWithVariableAsResource() {
...
-			).compile();
+				"try(resource){}}}"
+			).compile().getElements(new TypeFilter<>(CtTryWithResource.class)).get(0);
         List<CtResource<?>> resources = tryStmt.getResources();
         assertEquals(1, resources.size());
 		final CtResource<?> ctResource = resources.get(0);
-		assertTrue(ctResource instanceof CtVariableRead);
-        assertEquals("resource", ((CtVariableRead<?>) ctResource).getVariable().getSimpleName());
+		assertTrue(ctResource instanceof CtVariable);
+        assertEquals("resource", ((CtVariable<?>) ctResource).getSimpleName());
```
Comment (MartinWitt) [inline src/test/java/spoon/test/trycatch/TryCatchTest.java:433]: Convert the spaces to tabs here. We are not super strict with formatting in testcases but a method should either use spaces or tabs, not both

File: src/main/java/spoon/support/compiler/jdt/ParentExiter.java [hunk 3]
```diff
@@ -973,8 +976,33 @@ private CtCatch getLastCatcher(CtTry tryBlock) {
...
+			// special case of the resource being declared before
+			final CtVariableReference<?> variableRef = ((CtVariableRead<?>) child).getVariable();
+			if (variableRef.getDeclaration() != null) {
+				// getDeclaration works
+				tryWithResource.addResource((CtResource<?>) variableRef.getDeclaration().clone().setImplicit(true));
+			} else {
+				// we have to find it manually
+				for (ASTPair pair: this.jdtTreeBuilder.getContextBuilder().stack) {
+					final List<CtLocalVariable> variables = pair.element.getElements(new TypeFilter<>(CtLocalVariable.class));
+					if (variables.size() > 0) {
```
Comment (MartinWitt) [inline src/main/java/spoon/support/compiler/jdt/ParentExiter.java:992]:
```suggestion
					if (!variables.isEmpty()) {
```

File: src/main/java/spoon/support/compiler/jdt/ParentExiter.java [hunk 4]
```diff
@@ -973,8 +976,33 @@ private CtCatch getLastCatcher(CtTry tryBlock) {
...
 	@Override
 	public void visitCtTryWithResource(CtTryWithResource tryWithResource) {
-		if (child instanceof CtResource) {
-			tryWithResource.addResource((CtResource<?>) child);
+		if (child instanceof CtLocalVariable) {
+			// normal, happy path of declaring a new variable
+			tryWithResource.addResource((CtLocalVariable) child);
+		} else if (child instanceof CtVariableRead) {
+			// special case of the resource being declared before
+			final CtVariableReference<?> variableRef = ((CtVariableRead<?>) child).getVariable();
```
Comment (MartinWitt) [inline src/main/java/spoon/support/compiler/jdt/ParentExiter.java:984]: Are there any reasons for the final modifiers, especially for `Collections`?

**Review by MartinWitt — APPROVED** [review]
Only some very minor style suggestions, otherwise looks good. 👍

File: src/main/java/spoon/support/compiler/jdt/ParentExiter.java (hunk 4 above)
Comment (monperrus) [reply src/main/java/spoon/support/compiler/jdt/ParentExiter.java:984]: the IDE puts it automatically, and it does not hurt.

**Comment by monperrus** [comment]
Thanks fixed all comments

**Comment by MartinWitt** [comment]
thanks @monperrus

**Comment by monperrus** [comment]
I confirm that the regression identified in NpeFix is solved: https://ci.inria.fr/sos/job/npefix/1861/

**Pull request merged**